### PR TITLE
Magic Link Login: Mk 4

### LINKED ISSUE #1182: Magic Links Support
### Details:
- [x] Integrate the new Endpoints #1178
- [x] Implement support for the `simplenotemac://` callback URL #1179
- [x] Implement UI support for the Magic Links Flow #1180
- [x] Implement a new UI to be rendered during the Token Exchange Process #1181

### PR DESCRIPTION
**Pull request by jleandroperez**
### Details
In this PR we're wiring all of the Magic Link Authentication pieces together:

- Magic Link Login endpoints were integrated
- New SwiftUI views, to be presented while a Link is requested (and whenever a link ended up being invalid)

Closes #1182

### Test
1. Fresh install Simplenote macOS
2. Click on the `Log In` button
3. Enter your email
4. Click on `Instantly Log In with Email`

- [ ] Verify that, on completion, you get a UI indicating that you should check your email
- [ ] Verify that clicking on the `Log into Simplenote` button (on the Authentication Email) causes Simplenote to authenticate
- [ ] Verify that while the Auth process is executed, a spinner shows up onscreen with the legend `Logging in...`

### Test
1. Fresh install Simplenote macOS
2. Click on the same Authentication Link you've used above

- [ ] Verify that Simplenote displays the `Logging In...` legend
- [ ] Verify that shortly after, a UI shows up onscreen indicating that the link is no longer valid
- [ ] Verify that clicking on `Go back` causes the Auth UI to show up

### Release
`RELEASE-NOTES.txt` was updated in 3a1278a6 with:

> New Magic Link Authentication support

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by jleandroperez** [comment]
@charliescheer @roundhill Universal Links on the mac should be good now!!

P.s.: You **may need** to update your provisioning profiles!

File: Simplenote/MagicLinkConfirmationView.swift [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+// MARK: - MagicLinkConfirmationView
+//
+struct MagicLinkConfirmationView: View {
+    
+    @State var displaysInvalidLink = false
+    
+    var onDismissRequest: (() -> Void)?
+    
+    var body: some View {
+        VStack(spacing: 20) {
```
Comment (charliescheer) [inline Simplenote/MagicLinkConfirmationView.swift]: Nitpick: You got a lot of inline constants in this file

File: Simplenote/MagicLinkRequestedView.swift [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    private var emailLengthLimited: String {
+        guard email.count > Metrics.maximumEmailLength else {
+            return email
+        }
+        
+        return String(email.prefix(Metrics.maximumEmailLength)) + "..."
+    }
+
+    var body: some View {
+        VStack(alignment: .center, spacing: 10) {
```
Comment (charliescheer) [inline Simplenote/MagicLinkRequestedView.swift]: Nitpick: In line constants

File: Simplenote/MagicLinkRequestedView.swift [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+                .font(.title)
+                .padding(.bottom, Metrics.titlePaddingBottom)
+                        
+            Text("If an account exists, we've sent an email with a link that'll log you in to **\(emailLengthLimited)**")
+                .font(.title3)
+                .lineLimit(nil)
+                .multilineTextAlignment(.center)
+                .padding()
+            
+            SwiftUI.Button(action: switchToAuthenticationUI) {
```
Comment (charliescheer) [inline Simplenote/MagicLinkRequestedView.swift:48]: Do you need `SwiftUI.Button`?  Can't it just be `Button`?

**Review by charliescheer — APPROVED** [review]
looks good.  A few nitpicks

File: Simplenote/MagicLinkRequestedView.swift (hunk 3 above)
Comment (jleandroperez) [reply Simplenote/MagicLinkRequestedView.swift:48]: Unfortunately, there's another class named `Button`, so both collide 💔

File: Simplenote/MagicLinkConfirmationView.swift (hunk 1 above)
Comment (jleandroperez) [reply Simplenote/MagicLinkConfirmationView.swift]: Addressed, TY!

**Comment by jleandroperez** [comment]
Thank you Charlie!!

**Pull request merged**